loglist/diff_check_test.go

```diff
@@ -0,0 +1,222 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package loglist
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/mohae/deepcopy"
+)
+
+func pprint(stringList []string) string {
+	var pretty string
+	if buf, err := json.MarshalIndent(stringList, "", "  "); err == nil {
+		pretty = string(buf)
+	} else {
+		pretty = fmt.Sprintf("%v", stringList)
+	}
+	return pretty
+}
+
+func TestCheckOperatorsDiff(t *testing.T) {
+	tests := []struct {
+		name         string
+		branch       LogList
+		wantWarnings []string
+	}{
+		{
+			name: "Equal",
+			branch: LogList{
+				Operators: []Operator{
+					{ID: 0, Name: "Google"},
+					{ID: 1, Name: "Bob's CT Log Shop"},
+				}, Logs: []Log{},
+			},
+			wantWarnings: []string{},
+		},
+		{
+			name: "ShuffledRenamed",
+			branch: LogList{
+				Operators: []Operator{
+					{ID: 1, Name: "Bob's CT Log Shop+"},
+					{ID: 0, Name: "Google"},
+				}, Logs: []Log{},
+			},
+			wantWarnings: []string{},
+		},
+		{
+			name: "Missing",
+			branch: LogList{
+				Operators: []Operator{
+					{ID: 1, Name: "Bob's CT Log Shop"},
+				}, Logs: []Log{},
+			},
+			wantWarnings: []string{"Operator \"Google\" id=0 present at master log list but missing at branch."},
+		},
+		{
+			name: "Added",
+			branch: LogList{
+				Operators: []Operator{
+					{ID: 0, Name: "Google"},
+					{ID: 1, Name: "Bob's CT Log Shop"},
+					{ID: 2, Name: "Alice's CT Log Shop"},
+				}, Logs: []Log{},
+			},
+			wantWarnings: []string{},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			wl := warningList{warnings: []string{}}
+			checkMasterOpsMatchBranch(&sampleLogList, &test.branch, &wl)
+			printMismatchIfAny(t, wl.warnings, test.wantWarnings, "checkOperators:")
+		})
+	}
+}
+
+func TestCheckLogPairEquivalence(t *testing.T) {
+	type logPair struct {
+		log1         Log
+		log2         Log
+		wantWarnings []string
+	}
+	var tests = make(map[string]*logPair)
+	tests["Equal"] = &logPair{
+		log1:         deepcopy.Copy(sampleLogList.Logs[0]).(Log),
+		log2:         deepcopy.Copy(sampleLogList.Logs[0]).(Log),
+		wantWarnings: []string{},
+	}
+	tests["KeyURLMismatch"] = &logPair{
+		log1: deepcopy.Copy(sampleLogList.Logs[0]).(Log),
+		log2: deepcopy.Copy(sampleLogList.Logs[0]).(Log),
+		wantWarnings: []string{
+			"Log \"Google 'Aviator' log\" and log \"Google 'Aviator' log\" have different keys.",
+			"URL mismatch for logs \"Google 'Aviator' log\" and \"Google 'Aviator' log\": " +
+				"ct.googleapis.com/aviator/ != ct.googleapis.com/icarus/.",
```

Comment:
It actually makes the 'Contains' checks as suggested earlier. Please, refer to 'printMismatchIfAny' func. Left the full warnings as it's a test. To represent potential warnings.

Reply:
Yup I did see that after I wrote this comment :)  The point with using key word substrings instead of full warning strings is that if you were to tweak the wording of the warning strings in the library code, it would break the tests, even though the thing the tests are meant to be testing is still working correctly.  It means if you every change the warning strings generated by diff_check, you also have to change the tests, meaning the tests are fragile.

It's your code though, so your call, I won't press this point any further :)